### ISSUE TITLE
Categorical features for Apache Spark

### ISSUE THREAD

**Issue by ciromanfredi** [catboost/catboost#1998]
hi everyone I would like to know if there is a tutorial for the version of catboost for apache that manages categorical variables or in any case have an explanation of how it works... 
possible that there is no parameter that indicates categorical variables? 
Thanks also an example is fine

**Comment by andrey-khropov** [comment]
Spark ML library API is designed in a way that usually all features are stored in a `DataFrame` in a column with [`org.apache.spark.mllib.linalg.Vector` type](https://spark.apache.org/docs/3.1.1/mllib-data-types.html#local-vector).

CatBoost for Apache Spark follows that convention.

`Vector` stores only double values. So categorical features have to be transformed in a way that their values are remapped to consecutive integers (`0`, `1`, `2` ...) for each feature and then saved in these feature `Vector`s as doubles. 

If original categorical features values are not represented in such a way Spark ML Lib contains transformers for that: [StringIndexer](https://spark.apache.org/docs/3.1.1/ml-features.html#stringindexer) and [VectorIndexer](https://spark.apache.org/docs/3.1.1/ml-features.html#vectorindexer). After separate columns are transformed using `StringIndexer` then they can be merged to one `Vector` column with features using [VectorAssembler](https://spark.apache.org/docs/3.1.1/ml-features.html#vectorassembler).

Metadata about which features in the single `Vector` features column are categorical is stored in the special `ml_attr` attribute in column's metadata in `DataFrame`'s schema. It is set automatically by the aforementioned transformers but you can also set in explicitly as it is done in CatBoost for Apache Spark tests: [Scala example](https://github.com/catboost/catboost/blob/313d5a014b000933feecd29f214220738cf64d85/catboost/spark/catboost4j-spark/core/src/test/scala/ai/catboost/spark/PoolTestHelpers.scala#L14-L41), [PySpark example (analogous to Scala code)](https://github.com/catboost/catboost/blob/313d5a014b000933feecd29f214220738cf64d85/catboost/spark/catboost4j-spark/core/src/test/python/pool_test_helpers.py#L11-L46).

[Example in the tests that sets metadata explicitly](https://github.com/catboost/catboost/blob/313d5a014b000933feecd29f214220738cf64d85/catboost/spark/catboost4j-spark/core/src/test/scala/ai/catboost/spark/CatBoostClassifierTest.scala#L1172).

[Example in the tests that uses StringIndexer and VectorAssembler](https://github.com/catboost/catboost/blob/313d5a014b000933feecd29f214220738cf64d85/catboost/spark/catboost4j-spark/core/src/test/python/catboost_regressor_test.py#L501).

We'll defininitely add more info about this to our documentation.

**Comment by alimkidar** [comment]
Do we have the code for pyspark?

**Comment by andrey-khropov** [comment]
> Do we have the code example for pyspark?

Yes. See [here, in the tests](https://github.com/catboost/catboost/blob/313d5a014b000933feecd29f214220738cf64d85/catboost/spark/catboost4j-spark/core/src/test/python/catboost_regressor_test.py#L501)

**Comment by alimkidar** [comment]
> Yes. See [here, in the tests](https://github.com/catboost/catboost/blob/313d5a014b000933feecd29f214220738cf64d85/catboost/spark/catboost4j-spark/core/src/test/python/catboost_regressor_test.py#L501), it is mentioned in the comment above.

The StringIndexer will map the categorical feature into the index. But how do we tell the model those features are categorical features?  In the Python version, we can specify them on Pool method.

By the way, is it possible to use Target Encoding (CTRs) in spark, instead of One-Hot Encoding?

**Comment by andrey-khropov** [comment]
> > Yes. See [here, in the tests](https://github.com/catboost/catboost/blob/313d5a014b000933feecd29f214220738cf64d85/catboost/spark/catboost4j-spark/core/src/test/python/catboost_regressor_test.py#L501), it is mentioned in the comment above.
> 
> The StringIndexer will map the categorical feature into the index. But how do we tell the model those features are categorical features? In the Python version, we can specify them on Pool method.

As Spark ML library API has a built-in concept of categorical features stored in `ml_attr` attribute, CatBoost for Spark uses it. so specifying categorical features explicitly as in python (non-Spark) API is not necessary.

> By the way, is it possible to use Target Encoding (CTRs) in spark, instead of One-Hot Encoding?

Yes, but there's a limitation that CTR combinations with multiple features are not used because it is hard to compute them in a distributed manner effectively right now. It is like setting [`max_ctr_complexity` training parameter](https://catboost.ai/docs/en/references/training-parameters/ctr#max_ctr_complexity) to 1
